scripts/stop

```diff
@@ -3,13 +3,93 @@
 set -e
 projectRoot="$(a="/$0"; a=${a%/*}; a=${a:-.}; a=${a#/}/; cd "$a/.." || return; pwd)"
 
+# Set expected prefix for Docker resource names managed by this script
+dockerPrefix="ffc-demo-claim"
+
+show_help() {
+  echo "
+Stop service containers.
+
+Usage: scripts/stop [OPTION...] [-- [DOCKER_OPTION...]]
+
+Options:
+  -c, --clean  remove images and volumes as well as containers
+  -h, --help   display this help text
+  -l, --link   use this option when services were started with the '--link' flag
+               (uses the docker-compose.link.yaml override configuration)
+
+Docker options:
+  Any arguments after '--' are passed through to 'docker-compose down'.
+"
+}
+
+while :; do
+  case $1 in
+    -c|--clean)
+      clean="true"
+      ;;
+
+    -h|--help)
+      show_help
+      exit 0
+      ;;
+
+    -l|--link)
+      link="true"
+      ;;
+
+    --)
+      shift
+      break
+      ;;
+
+    -?*)
+      echo "Unrecognised argument: $1"
+      show_help
+      exit 1
+      ;;
+
+    *)
+      break
+  esac
+
+  shift
+done
+
+compose() {
+  if [ "${link}" = "true" ]; then
+    docker-compose -f docker-compose.yaml -f docker-compose.link.yaml $@
+  else
+    docker-compose $@
+  fi
+}
+
 (
   cd "${projectRoot}"
 
-  existingServices="$( docker-compose ps -q )"
-  existingVolumes="$( docker volume ls --filter name=^ffc-demo-claim-service --format={{.Name}} )"
+  existingContainers="$( docker-compose ps -q )"
+  existingVolumes="$( docker volume ls --filter name=^${dockerPrefix}-service --format={{.Name}} )"
+
+  # Only bring down services if there are existing containers/volumes
+  if [ -n "${existingContainers}${existingVolumes}" ]; then
+    if [ "${clean}" = "true" ]; then
+      compose down --remove-orphans --rmi=local --volumes $@
+    else
+      compose down $@
+    fi
+  else
+    echo "Services already stopped."
+  fi
 
-  if [ -n "${existingServices}${existingVolumes}" ]; then
-    docker-compose down -v
+  if [ "${clean}" = "true" ]; then
+    docker image prune -f
```

Comment:
This is a potenitally a long running task hidden within the stop script

Reply:
It can take a while if you have a lot of dangling images on your machine, although if you prune frequently it tends to be pretty quick.

This does clear out all dangling images on the machine (not just from this project), so goes beyond the intended scope of the script and is slower than it needs to be. I couldn't find a reliable way to remove only the images created by a single project because of limitations in the Docker CLI:

1. `docker-compose images` only lists images currently in use, not including past images
2. `docker image rm` with specific image IDs doesn't remove the dangling images left behind

I really wanted the `--clean` flag to remove old and current images for the project to put your system into a state as if the project had never been run.

Any ideas on how to properly clean out all images from a specific project without affecting other images?

Reply:
I guess it's unavoidable, maybe add a message saying clearing dangling images to inform the user it may take a while?

Reply:
Good idea, I've added a message.